Changes in module 'nf-core/panaroo/run'
--- modules/nf-core/panaroo/run/main.nf
+++ modules/nf-core/panaroo/run/main.nf
@@ -2,10 +2,10 @@
     tag "$meta.id"
     label 'process_medium'
 
-    conda "bioconda::panaroo=1.2.9"
+    conda (params.enable_conda ? "bioconda::panaroo=1.3.2" : null)
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/panaroo:1.2.9--pyhdfd78af_0':
-        'quay.io/biocontainers/panaroo:1.2.9--pyhdfd78af_0' }"
+        'https://depot.galaxyproject.org/singularity/panaroo:1.3.2--pyhdfd78af_0':
+        'quay.io/biocontainers/panaroo:1.3.2--pyhdfd78af_0' }"
 
     input:
     tuple val(meta), path(gff)
@@ -33,4 +33,15 @@
         panaroo: \$(echo \$(panaroo --version 2>&1) | sed 's/^.*panaroo //' ))
     END_VERSIONS
     """
+    stub:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    mkdir results
+    touch results/core_gene_alignment.aln
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        panaroo: \$(echo \$(panaroo --version 2>&1) | sed 's/^.*panaroo //' ))
+    END_VERSIONS
+    """
 }

************************************************************
