Changes in module 'nf-core/kraken2/kraken2'
--- modules/nf-core/kraken2/kraken2/main.nf
+++ modules/nf-core/kraken2/kraken2/main.nf
@@ -2,7 +2,7 @@
     tag "$meta.id"
     label 'process_high'
 
-    conda "bioconda::kraken2=2.1.2 conda-forge::pigz=2.6"
+    conda (params.enable_conda ? "bioconda::kraken2=2.1.2 conda-forge::pigz=2.6" : null)
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/mulled-v2-5799ab18b5fc681e75923b2450abaa969907ec98:87fc08d11968d081f3e8a37131c1f1f6715b6542-0' :
         'quay.io/biocontainers/mulled-v2-5799ab18b5fc681e75923b2450abaa969907ec98:87fc08d11968d081f3e8a37131c1f1f6715b6542-0' }"

************************************************************
