<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="Beiko Lab" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>ParallelEvolCCM - ARETE</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "ParallelEvolCCM";
        var mkdocs_page_input_path = "evolccm.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="..">
          <img src="../assets/arete-symbol.png" class="logo" alt="Logo"/>
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../usage/">Usage</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../faq/">FAQ</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../output/">Output</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../issues/">Known issues</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../CITATIONS/">Citations</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../ROADMAP/">Roadmap</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../contributing/">Contributing</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Reference</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../resource_profiles/">Dataset Size</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../params/">Parameters</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../subsampling/">Subsampling</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">ParallelEvolCCM</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#parallelevolccm-inputs">ParallelEvolCCM inputs</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#using-parallelevolccm-by-itself">Using ParallelEvolCCM by itself</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#using-parallelevolccm-with-arete">Using ParallelEvolCCM with ARETE</a>
    </li>
    </ul>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">ARETE</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">Reference</li>
      <li class="breadcrumb-item active">ParallelEvolCCM</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/beiko-lab/arete/edit/master/docs/evolccm.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="parallelevolccm-usage">ParallelEvolCCM usage</h1>
<p>ParallelEvolCCM is a tool for the identification of coordinated gain and loss of features.
The method is described in detail in the following publication:</p>
<ul>
<li><a href="https://doi.org/10.1093/sysbio/syac052">The Community Coevolution Model with Application to the Study of Evolutionary Relationships between Genes Based on Phylogenetic Profiles</a></li>
</ul>
<p>If you use ParallelEvolCCM in your analysis, please cite the above publication.</p>
<h2 id="parallelevolccm-inputs">ParallelEvolCCM inputs</h2>
<p>The ParallelEvolCCM tool requires two inputs:</p>
<ul>
<li>A phylogenetic tree in Newick format</li>
<li>A presence/absence table in TSV format.</li>
</ul>
<p>The presence/absence TSV must have genome names equal to the ones in the tree in a 'genome_id' column,
with all other columns representing features absent (0) or present (1) in each genome. I.e.:</p>
<pre><code>genome_id   plasmid_AA155   plasmid_AA161
ED010   0   0
ED017   0   1
ED040   0   0
ED073   0   1
ED075   1   1
ED082   0   1
ED142   0   1
ED178   0   1
ED180   0   0
</code></pre>
<h2 id="using-parallelevolccm-by-itself">Using ParallelEvolCCM by itself</h2>
<p>The ParallelEvolCCM tool is a command line tool written in R.
It is available through the <a href="https://github.com/beiko-lab/arete/blob/master/bin/ParallelEvolCCM.R">bin/ParallelEvolCCM.R</a> script.</p>
<p>To download the tool and make it executable, run:</p>
<pre><code class="language-bash">wget https://raw.githubusercontent.com/beiko-lab/arete/master/bin/ParallelEvolCCM.R
chmod +x ParallelEvolCCM.R
</code></pre>
<p>Then, ensure all EvolCCM dependencies are installed.
You can install them by running the following command in your R console:</p>
<pre><code class="language-r">install.packages(c('ape', 'dplyr', 'phytools', 'foreach', 'doParallel', 'gplots', 'remotes'))
remotes::install_github('beiko-lab/evolCCM')
</code></pre>
<p>You can then run the tool like this:</p>
<pre><code class="language-bash">./ParallelEvolCCM.R --intree tree.nwk --intable feature_table.tsv.gz --cores -1
</code></pre>
<ul>
<li><code>--intree</code> specifies the phylogenetic tree in Newick format.</li>
<li><code>--intable</code> specifies the feature table in compressed TSV format.</li>
<li><code>--cores</code> specifies the number of cores to use. Use <code>-1</code> to use all available cores.</li>
</ul>
<p>Additional parameters can be found by running <code>./ParallelEvolCCM.R</code> with no additional parameters.</p>
<h2 id="using-parallelevolccm-with-arete">Using ParallelEvolCCM with ARETE</h2>
<p>The ParallelEvolCCM tool is also made available through the <code>evolccm</code> entry in ARETE.
Making it possible to run the tool with Docker or Singularity.</p>
<p>To execute the ParallelEvolCCM tool with ARETE, run the following command:</p>
<pre><code class="language-bash">nextflow run beiko-lab/ARETE \
  -entry evolccm \
  --core_gene_tree core_gene_alignment.tre \
  --feature_profile feature_profile.tsv.gz \
  -profile docker
</code></pre>
<p>The parameters being:</p>
<ul>
<li><code>--core_gene_tree</code> - The reference tree, coming from a core genome alignment,
  like the one generated by the <code>phylo</code> entry in ARETE.</li>
<li><code>--feature_profile</code> - A presence/absence TSV matrix of features
  in genomes, like the one created in ARETE's <code>annotation</code> entry.</li>
<li><code>-profile</code> - The profile to use. In this case, <code>docker</code>.</li>
</ul>
<p>For more information, check the <a href="https://beiko-lab.github.io/arete/">full ARETE documentation</a>.</p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../subsampling/" class="btn btn-neutral float-left" title="Subsampling"><span class="icon icon-circle-arrow-left"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/beiko-lab/arete" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../subsampling/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
