Changes in module 'nf-core/fasttree'
--- modules/nf-core/fasttree/main.nf
+++ modules/nf-core/fasttree/main.nf
@@ -1,7 +1,7 @@
 process FASTTREE {
     label 'process_medium'
 
-    conda "bioconda::fasttree=2.1.10"
+    conda (params.enable_conda ? "bioconda::fasttree=2.1.10" : null)
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/fasttree:2.1.10--h516909a_4' :
         'quay.io/biocontainers/fasttree:2.1.10--h516909a_4' }"

************************************************************
