<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="Beiko Lab" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Citations - ARETE</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Citations";
        var mkdocs_page_input_path = "CITATIONS.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> ARETE
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../usage/">Usage</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../faq/">FAQ</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../output/">Output</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="./">Citations</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#nf-core">nf-core</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#nextflow">Nextflow</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#pipeline-tools">Pipeline tools</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#software-packagingcontainerisation-tools">Software packaging/containerisation tools</a>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../ROADMAP/">Roadmap</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Reference</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../params/">Parameters</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../subsampling/">Subsampling</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">ARETE</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a> &raquo;</li>
      <li class="breadcrumb-item active">Citations</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/beiko-lab/arete/edit/master/docs/CITATIONS.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="beiko-labarete-citations">beiko-lab/ARETE: Citations</h1>
<h2 id="nf-core"><a href="https://pubmed.ncbi.nlm.nih.gov/32055031/">nf-core</a></h2>
<blockquote>
<p>Ewels PA, Peltzer A, Fillinger S, Patel H, Alneberg J, Wilm A, Garcia MU, Di Tommaso P, Nahnsen S. The nf-core framework for community-curated bioinformatics pipelines. Nat Biotechnol. 2020 Mar;38(3):276-278. doi: 10.1038/s41587-020-0439-x. PubMed PMID: 32055031.</p>
</blockquote>
<h2 id="nextflow"><a href="https://pubmed.ncbi.nlm.nih.gov/28398311/">Nextflow</a></h2>
<blockquote>
<p>Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. Nextflow enables reproducible computational workflows. Nat Biotechnol. 2017 Apr 11;35(4):316-319. doi: 10.1038/nbt.3820. PubMed PMID: 28398311.</p>
</blockquote>
<h2 id="pipeline-tools">Pipeline tools</h2>
<ul>
<li><a href="https://github.com/Ecogenomics/CheckM">CheckM</a></li>
</ul>
<blockquote>
<p>Parks DH, Imelfort M, Skennerton CT, Hugenholtz P, Tyson GW. 2015. <a href="http://genome.cshlp.org/content/25/7/1043.short">CheckM: assessing the quality of microbial genomes recovered from isolates, single cells, and metagenomes</a>. Genome Research, 25: 1043–1055.</p>
</blockquote>
<ul>
<li><a href="https://github.com/bbuchfink/diamond">DIAMOND</a></li>
</ul>
<blockquote>
<p>Buchfink B, Xie C, Huson DH <a href="http://dx.doi.org/10.1038/nmeth.3176">Fast and sensitive protein alignment using DIAMOND.</a> <em>Nat. Methods.</em> 12, 59–60 (2015)</p>
</blockquote>
<ul>
<li>
<p><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></p>
</li>
<li>
<p><a href="https://www.ncbi.nlm.nih.gov/pubmed/30423086/">FastP</a></p>
</li>
</ul>
<blockquote>
<p>Chen S, Zhou Y, Chen Y, Gu J. fastp: an ultra-fast all-in-one FASTQ preprocessor. Bioinformatics. 2018 Sep 1;34(17):i884-i890. doi: 10.1093/bioinformatics/bty560. PubMed PMID: 30423086; PubMed Central PMCID: PMC6129281.</p>
</blockquote>
<ul>
<li><a href="https://doi.org/10.1093/molbev/msp077">FastTree</a></li>
</ul>
<blockquote>
<p>Morgan N. Price, Paramvir S. Dehal, Adam P. Arkin, FastTree: Computing Large Minimum Evolution Trees with Profiles instead of a Distance Matrix, Molecular Biology and Evolution, Volume 26, Issue 7, July 2009, Pages 1641–1650, https://doi.org/10.1093/molbev/msp077</p>
</blockquote>
<ul>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/32011700/">IQ-TREE2</a></li>
</ul>
<blockquote>
<p>Minh BQ, Schmidt HA, Chernomor O, Schrempf D, Woodhams MD, von Haeseler A, Lanfear R. IQ-TREE 2: New Models and Efficient Methods for Phylogenetic Inference in the Genomic Era. Mol Biol Evol. 2020 May 1;37(5):1530-1534. doi: 10.1093/molbev/msaa015. Erratum in: Mol Biol Evol. 2020 Aug 1;37(8):2461. PMID: 32011700; PMCID: PMC7182206.</p>
</blockquote>
<ul>
<li><a href="https://doi.org/10.1186/s13059-019-1891-0">Kraken2</a></li>
</ul>
<blockquote>
<p>Wood, D et al., 2019. Improved metagenomic analysis with Kraken 2. Genome Biology volume 20, Article number: 257. doi: 10.1186/s13059-019-1891-0.</p>
</blockquote>
<ul>
<li><a href="https://github.com/phac-nml/mob-suite">MOB-SUITE</a></li>
</ul>
<blockquote>
<p>Robertson, James, and John H E Nash. “MOB-suite: software tools for clustering, reconstruction and typing of plasmids from draft assemblies.” Microbial genomics vol. 4,8 (2018): e000206. doi:10.1099/mgen.0.000206
Robertson, James et al. “Universal whole-sequence-based plasmid typing and its utility to prediction of host range and epidemiological surveillance.” Microbial genomics vol. 6,10 (2020): mgen000435. doi:10.1099/mgen.0.000435</p>
</blockquote>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/pubmed/27312411/">MultiQC</a></li>
</ul>
<blockquote>
<p>Ewels P, Magnusson M, Lundin S, Käller M. MultiQC: summarize analysis results for multiple tools and samples in a single report. Bioinformatics. 2016 Oct 1;32(19):3047-8. doi: 10.1093/bioinformatics/btw354. Epub 2016 Jun 16. PubMed PMID: 27312411; PubMed Central PMCID: PMC5039924.</p>
</blockquote>
<ul>
<li><a href="https://doi.org/10.1099/mgen.0.000685">Bakta</a></li>
</ul>
<blockquote>
<p>Schwengers O., Jelonek L., Dieckmann M. A., Beyvers S., Blom J., Goesmann A. (2021). Bakta: rapid and standardized annotation of bacterial genomes via alignment-free sequence identification. Microbial Genomics, 7(11). https://doi.org/10.1099/mgen.0.000685</p>
</blockquote>
<ul>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/24642063/">Prokka</a></li>
</ul>
<blockquote>
<p>Seemann T. Prokka: rapid prokaryotic genome annotation. Bioinformatics. 2014 Jul 15;30(14):2068-9. doi: 10.1093/bioinformatics/btu153. Epub 2014 Mar 18. PMID: 24642063.</p>
</blockquote>
<ul>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/23422339/">QUAST</a></li>
</ul>
<blockquote>
<p>Gurevich A, Saveliev V, Vyahhi N, Tesler G. QUAST: quality assessment tool for genome assemblies. Bioinformatics. 2013 Apr 15;29(8):1072-5. doi: 10.1093/bioinformatics/btt086. Epub 2013 Feb 19. PMID: 23422339; PMCID: PMC3624806.</p>
</blockquote>
<ul>
<li><a href="https://github.com/arpcard/rgi">RGI</a></li>
</ul>
<blockquote>
<p>Alcock et al. 2020. CARD 2020: antibiotic resistome surveillance with the comprehensive antibiotic resistance database. Nucleic Acids Research, Volume 48, Issue D1, Pages D517-525 [PMID 31665441]</p>
</blockquote>
<ul>
<li><a href="https://doi.org/10.3390/microorganisms10040700">IntegronFinder</a></li>
</ul>
<blockquote>
<p>Néron, Bertrand, Eloi Littner, Matthieu Haudiquet, Amandine Perrin, Jean Cury, and Eduardo P.C. Rocha. 2022. IntegronFinder 2.0: Identification and Analysis of Integrons across Bacteria, with a Focus on Antibiotic Resistance in Klebsiella Microorganisms 10, no. 4: 700. https://doi.org/10.3390/microorganisms10040700</p>
</blockquote>
<ul>
<li><a href="https://github.com/gtonkinhill/panaroo">Panaroo</a></li>
</ul>
<blockquote>
<p>Tonkin-Hill, G., MacAlasdair, N., Ruis, C. et al. Producing polished prokaryotic pangenomes with the Panaroo pipeline. Genome Biol 21, 180 (2020). https://doi.org/10.1186/s13059-020-02090-4</p>
</blockquote>
<ul>
<li><a href="https://doi.org/10.1371/journal.pcbi.1007732">PPanGGoLiN</a></li>
</ul>
<blockquote>
<p>Gautreau G et al. (2020) PPanGGOLiN: Depicting microbial diversity via a partitioned pangenome graph. PLOS Computational Biology 16(3): e1007732. https://doi.org/10.1371/journal.pcbi.1007732</p>
</blockquote>
<ul>
<li><a href="https://doi.org/10.1101/gr.241455.118">PopPUNK</a></li>
</ul>
<blockquote>
<p>Lees JA, Harris SR, Tonkin-Hill G, Gladstone RA, Lo SW, Weiser JN, Corander J, Bentley SD, Croucher NJ. Fast and flexible bacterial genomic epidemiology with PopPUNK. Genome Res. 2019 Feb;29(2):304-316. doi: 10.1101/gr.241455.118. Epub 2019 Jan 24. PMID: 30679308; PMCID: PMC6360808.</p>
</blockquote>
<ul>
<li><a href="https://doi.org/10.1101/453142">SKA2</a></li>
</ul>
<blockquote>
<p>Harris SR. 2018. SKA: Split Kmer Analysis Toolkit for Bacterial Genomic Epidemiology. bioRxiv 453142 doi: https://doi.org/10.1101/453142</p>
</blockquote>
<ul>
<li><a href="https://doi.org/10.1093/nar/gku1196">Gubbins</a></li>
</ul>
<blockquote>
<p>Croucher N. J., Page A. J., Connor T. R., Delaney A. J., Keane J. A., Bentley S. D., Parkhill J., Harris S.R. "Rapid phylogenetic analysis of large samples of recombinant bacterial whole genome sequences using Gubbins". doi:10.1093/nar/gku1196, Nucleic Acids Research, 2014.</p>
</blockquote>
<ul>
<li>
<p><a href="https://github.com/rrwick/Verticall">Verticall</a></p>
</li>
<li>
<p><a href="https://pubmed.ncbi.nlm.nih.gov/28348851/">SNP-sites</a></p>
</li>
</ul>
<blockquote>
<p>Page AJ, Taylor B, Delaney AJ, Soares J, Seemann T, Keane JA, Harris SR. SNP-sites: rapid efficient extraction of SNPs from multi-FASTA alignments. Microb Genom. 2016 Apr 29;2(4):e000056. doi: 10.1099/mgen.0.000056. PMID: 28348851; PMCID: PMC5320690.</p>
</blockquote>
<ul>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/28594827/">Unicycler</a></li>
</ul>
<blockquote>
<p>Wick RR, Judd LM, Gorrie CL, Holt KE. Unicycler: Resolving bacterial genome assemblies from short and long sequencing reads. PLoS Comput Biol. 2017 Jun 8;13(6):e1005595. doi: 10.1371/journal.pcbi.1005595. PMID: 28594827; PMCID: PMC5481147.</p>
</blockquote>
<ul>
<li><a href="https://doi.org/10.1093/bioinformatics/bty095">IslandPath</a></li>
</ul>
<blockquote>
<p>Claire Bertelli, Fiona S L Brinkman, Improved genomic island predictions with IslandPath-DIMOB, Bioinformatics, Volume 34, Issue 13, 01 July 2018, Pages 2161–2167, https://doi.org/10.1093/bioinformatics/bty095</p>
</blockquote>
<ul>
<li><a href="https://doi.org/10.1093/nar/gks406">PhiSpy</a><blockquote>
<p>Sajia Akhter, Ramy K. Aziz, Robert A. Edwards; PhiSpy: a novel algorithm for finding prophages in bacterial genomes that combines similarity- and composition-based strategies. Nucl Acids Res 2012; 40 (16): e126. doi: 10.1093/nar/gks406</p>
</blockquote>
</li>
</ul>
<h2 id="software-packagingcontainerisation-tools">Software packaging/containerisation tools</h2>
<ul>
<li><a href="https://anaconda.com">Anaconda</a></li>
</ul>
<blockquote>
<p>Anaconda Software Distribution. Computer software. Vers. 2-2.4.0. Anaconda, Nov. 2016. Web.</p>
</blockquote>
<ul>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/29967506/">Bioconda</a></li>
</ul>
<blockquote>
<p>Grüning B, Dale R, Sjödin A, Chapman BA, Rowe J, Tomkins-Tinch CH, Valieris R, Köster J; Bioconda Team. Bioconda: sustainable and comprehensive software distribution for the life sciences. Nat Methods. 2018 Jul;15(7):475-476. doi: 10.1038/s41592-018-0046-7. PubMed PMID: 29967506.</p>
</blockquote>
<ul>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/28379341/">BioContainers</a></li>
</ul>
<blockquote>
<p>da Veiga Leprevost F, Grüning B, Aflitos SA, Röst HL, Uszkoreit J, Barsnes H, Vaudel M, Moreno P, Gatto L, Weber J, Bai M, Jimenez RC, Sachsenberg T, Pfeuffer J, Alvarez RV, Griss J, Nesvizhskii AI, Perez-Riverol Y. BioContainers: an open-source and community-driven framework for software standardization. Bioinformatics. 2017 Aug 15;33(16):2580-2582. doi: 10.1093/bioinformatics/btx192. PubMed PMID: 28379341; PubMed Central PMCID: PMC5870671.</p>
</blockquote>
<ul>
<li><a href="https://dl.acm.org/doi/10.5555/2600239.2600241">Docker</a></li>
</ul>
<blockquote>
<p>Dirk Merkel. 2014. Docker: lightweight Linux containers for consistent development and deployment. Linux J. 2014, 239, Article 2 (March 2014).</p>
</blockquote>
<ul>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/28494014/">Singularity</a><blockquote>
<p>Kurtzer GM, Sochat V, Bauer MW. Singularity: Scientific containers for mobility of compute. PLoS One. 2017 May 11;12(5):e0177459. doi: 10.1371/journal.pone.0177459. eCollection 2017. PubMed PMID: 28494014; PubMed Central PMCID: PMC5426675.</p>
</blockquote>
</li>
</ul>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../output/" class="btn btn-neutral float-left" title="Output"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../ROADMAP/" class="btn btn-neutral float-right" title="Roadmap">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/beiko-lab/arete" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../output/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../ROADMAP/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
