Changes in module 'nf-core/prokka'
--- modules/nf-core/prokka/main.nf
+++ modules/nf-core/prokka/main.nf
@@ -2,7 +2,7 @@
     tag "$meta.id"
     label 'process_low'
 
-    conda "bioconda::prokka=1.14.6"
+    conda (params.enable_conda ? "bioconda::prokka=1.14.6" : null)
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/prokka%3A1.14.6--pl5321hdfd78af_4' :
         'quay.io/biocontainers/prokka:1.14.6--pl5321hdfd78af_4' }"

************************************************************
