Changes in module 'nf-core/quast'
--- modules/nf-core/quast/main.nf
+++ modules/nf-core/quast/main.nf
@@ -1,7 +1,7 @@
 process QUAST {
     label 'process_medium'
 
-    conda "bioconda::quast=5.2.0"
+    conda (params.enable_conda ? "bioconda::quast=5.2.0" : null)
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/quast:5.2.0--py39pl5321h2add14b_1' :
         'quay.io/biocontainers/quast:5.2.0--py39pl5321h2add14b_1' }"

************************************************************
