Changes in module 'nf-core/quast'
--- modules/nf-core/quast/main.nf
+++ modules/nf-core/quast/main.nf
@@ -1,7 +1,7 @@
 process QUAST {
     label 'process_medium'
 
-    conda "bioconda::quast=5.2.0"
+    conda (params.enable_conda ? "bioconda::quast=5.2.0" : null)
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/quast:5.2.0--py39pl5321h2add14b_1' :
         'quay.io/biocontainers/quast:5.2.0--py39pl5321h2add14b_1' }"
@@ -15,7 +15,8 @@
 
     output:
     path "${prefix}"    , emit: results
-    path '*.tsv'        , emit: tsv
+    path "*.tsv"        , emit: tsv
+    path "${prefix}/transposed_report.tsv", emit: transposed_report
     path "versions.yml" , emit: versions
 
     when:
@@ -33,7 +34,7 @@
         $features \\
         --threads $task.cpus \\
         $args \\
-        ${consensus.join(' ')}
+        \$(cat ${consensus.join(' ')})
 
     ln -s ${prefix}/report.tsv
 
@@ -42,4 +43,16 @@
         quast: \$(quast.py --version 2>&1 | sed 's/^.*QUAST v//; s/ .*\$//')
     END_VERSIONS
     """
+    stub:
+    prefix   = task.ext.prefix ?: 'quast'
+    """
+    mkdir ${prefix}
+    touch ${prefix}/report.tsv
+    touch ${prefix}/transposed_report.tsv
+    touch ${prefix}_quast_stub.tsv
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        quast: \$(quast.py --version 2>&1 | sed 's/^.*QUAST v//; s/ .*\$//')
+    END_VERSIONS
+    """
 }

************************************************************
